### PR TITLE
[gitlab] Disable repo permissions syncing if auth type is oauth

### PR DESCRIPTION
**Pull request by pjlast**
Disables repo-based permissions syncing if the auth provider type for a GitLab code host connection is set as oauth. Repo-based permissions syncing has a lot of gotchas when it comes to token permissions, and they are _very_ hard to debug and fundamentally flawed.

## Test plan

Updated unit tests

### CODE REVIEW

File: enterprise/internal/authz/gitlab/oauth.go [hunk 1]
```diff
@@ -128,20 +128,5 @@ func (p *OAuthProvider) FetchUserPerms(ctx context.Context, account *extsvc.Acco
...
-	var client *gitlab.Client
-	switch p.tokenType {
-	case gitlab.TokenTypeOAuth:
-		client = p.clientProvider.GetOAuthClient(p.token)
-	default:
-		client = p.clientProvider.GetPATClient(p.token, "")
-	}
-
-	return listMembers(ctx, client, repo.ID)
+	return nil, authz.ErrUnimplemented{}
```
Comment (sashaostrikov) [inline enterprise/internal/authz/gitlab/oauth.go:131]: Let's add a doc about why is it not implemented

**Review by sashaostrikov — APPROVED** [review]

File: enterprise/internal/authz/gitlab/oauth.go (hunk 1 above)
Comment (sashaostrikov) [reply enterprise/internal/authz/gitlab/oauth.go:131]: I think maybe we should also update the docs page 🤔

Comment (pjlast) [reply enterprise/internal/authz/gitlab/oauth.go:131]: Can't add a changelog entry before the PR exists 🙃

Comment (sashaostrikov) [reply enterprise/internal/authz/gitlab/oauth.go:131]: true. that's why I end up with a changelog entry with a `[TODO](TODO)` _which I forget to update in 80% cases_

File: CHANGELOG.md [hunk 2]
```diff
@@ -39,6 +39,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Removed an incorrect beta label on the Search Context creation page [#51188](https://github.com/sourcegraph/sourcegraph/pull/51188)
 - Multi-version upgrades to version `5.0.2` in a fully airgapped environment will not work without the command `--skip-drift-check`. [#51164](https://github.com/sourcegraph/sourcegraph/pull/51164)
 - Could not set "permissions.syncOldestUsers" or "permissions.syncOldestRepos" to zero. [#51255](https://github.com/sourcegraph/sourcegraph/pull/51255)
+- GitLab code host connections will no longer use Personal Access Tokens to sync repository permissions when the authentication provider is set as "oauth". This prevents inconsistent user permissions from occuring. [#51452](https://github.com/sourcegraph/sourcegraph/pull/51452)
```
Comment (kopancek) [inline CHANGELOG.md:42]:
```suggestion
- GitLab code host connections will disable repo-centric repository permission syncs when the authentication provider is set as "oauth". This prevents repo-centric permission sync from getting incorrect data. [#51452](https://github.com/sourcegraph/sourcegraph/pull/51452)
```

File: doc/admin/external_service/gitlab.md [hunk 3]
```diff
@@ -88,6 +88,9 @@ Then, [add or edit a GitLab connection](#repository-syncing) and include the `au
 }
 ```
 
+In this case, a user's OAuth token will be used to verify that user's access to repositories.
```
Comment (kopancek) [inline doc/admin/external_service/gitlab.md:91]:
```suggestion
In this case, a user's OAuth token will be used to get a list of repositories that the user can access.
```

**Review by kopancek — COMMENTED** [review]
Suggested some wording changes

**Pull request merged**